src/twisted/protocols/_smb/base.py

```diff
@@ -0,0 +1,207 @@
+        """
+        p = [self._values.get(n, self.zero(n)) for n in self._fields_names]
+        return struct.pack(self._pattern, *p) + self.buffer
+
+    def unpack(self, data):
+        """
+        unpack data according to the pattern
+        exceas data goes in buffer
+
+        @param data: the data
+        @type data: L{bytes}
+        """
+        ret = struct.unpack(self._pattern, data[:self._size])
+        for i in range(len(ret)):
+            self._values[self._fields_names[i]] = ret[i]
+        self.buffer = data[self._size:]
+
+    def clear(self):
+        """
+        clear internal fields
+        """
+        self._values = {}
+        self.buffer = b''
+
+    def __len__(self):
+        return self._size
+
+    def __bytes__(self):
+        return self.pack()
+
+
+
+def nstruct(fields):
+    """
+    dynamically subclass GeneralStruct with a speeific binary structure
+
+    @param fields: space-separated list field:type.
+    Type is a single char as used in L{struct.pack}
+    (but only use s and numeric types)
+    @type fields: L{str}
+    """
+    d = {}
+    d['_fields_names'] = []
+    d['_types'] = []
+    for f in fields.split():
+        f = f.split(":")
+        d['_fields_names'].append(f[0])
+        d['_types'].append(f[1])
+    d['_pattern'] = "<" + "".join(d['_types'])
+    d['_size'] = struct.calcsize(d['_pattern'])
+    return type(d['_pattern'], (GeneralStruct, ), d)
+
+
+
+def int32key(d, val):
+    """
+    generate a new random key for a dictionary
+    @param d: dictionary with 32-bit int keys
+    @type d: L{dict}
+    @param val: new dictionary value
+    @rtype: L{int}
+    """
+    assert len(d) < 0xc0000000  # otherwiae dict so big hard to find keys
+    n, = struct.unpack("<I", secureRandom(4, True))
+    while n == 0 or n in d:
+        n, = struct.unpack("<I", secureRandom(4, True))
+    d[n] = val
+    return n
+
+
+
+class SMBPacketReceiver(protocol.Protocol):
+    """
+    basic SMB 2.0 packets over TCP have a 4-byte header: null byte
+    and 24-bit length field
+    this base class processes these headers
+    """
+    def __init__(self):
+        self.data = b''
+
+    def dataReceived(self, data):
+        self.data += data
+        self._processData()
+
+    def _processData(self):
+        if len(self.data) < 5:
+            return
+        x, y = struct.unpack("!xBH", self.data[:4])
+        size = (x << 16) + y
+        if len(self.data) < size + 4:
+            return
+        self.packetReceived(self.data[4:4 + size])
+        self.data = self.data[4 + size:]
+        self._processData()
+
+    def sendPacket(self, data):
+        """
+        send data with 4 byte header
+
+        @param dara: packet to send
+        @type data: L{bytes}
+        """
+        size = len(data)
+        assert size < 0xffffff
+        x = (size & 0xff0000) >> 16
+        y = size & 0xffff
+        self.transport.write(struct.pack("!BBH", 0, x, y) + data)
+
+    def packetReceived(self, packet):
```

Comment:
I know this stack-of-superclasses is a pattern throughout twisted but it's really bad 😬.  We should have a `self._packetReceiver` object and this should be `self._packetReceiver.packetReceived` inside `dataReceived`, then delegate on down for each level of transformation.  Avoiding subclassing as an interface avoids lots of namespace-collision problems down the road (and allows for a super narrow top-level interface that doesn't expose any of these internals.

Reply:
how does the contained object (PacketReceiver) call out to the container with new incoming packets? Deferred doesn't work because it's for multiple use. Does the container supply a callback ?

Reply:
`core.py` has been moved to a composition based style and I've removed the `SMBConnection`